broadinstitute/single_cell_portal_core
# Refactoring explore tab logic (SCP-3167)

This implements the suggestions from Jerome and the last couple of demos.
  1.  It consolidates the former 'cluster' 'scatter' and 'spatial' plots into a  single "scatter" tab.  This is both good from a UX perspective, and very important from a performance perspective, as it makes it much easier to avoid replots of cluster plots when going back and forth between views
  2. It adds a 'loading' tab for while we're waiting for the exploreInfo to load
  3. It hides controls in the 'view options' that are not relevant to the currently shown tab
  4. It moves the view options controls to be part of the tab with the white background
![image](https://user-images.githubusercontent.com/2800795/113235907-da845f00-9271-11eb-855d-30cedacc286f.png)

For the code, it simplifies ExploreView -- that component is now just responsible for fetching the explore data nad managing defaults.  ExploreDisplayTabs picks up the logic that used to be in ExploreView.  ScatterTab now has the logic for rendering all the different scatter plot configurations.  This logic has also been moved outside of JSX and into pure javascript so it's easier to follow.  A bunch of logic has been added to manage webgl contexts in a more sophisticated way than we were before.

To test:
 1.  Load the 'chicken raw counts' study with react_explore turned on
 2. Observe the 'loading' tab briefly appears before the cluster.
 3. Observe that once the cluster plot shows, the tab is labeled 'scatter', and observe that the distribution and heatmap controls are hidden
 4. Search for gene 'foo'
 5. Observe the tab is still called 'scatter', and that the cluster graph did not need to re-render
 6. go to the distribution tab, and observe that the 'scatter' controls are replaced with 'distribution' controls
 7. go back to the scatter tab, and select two spatial groups
 8. Observe the spatial plots appear, and the existing scatter plots do not need to be redrawn
 9. Search for the gene 'bar'
 10. Observe the gene expression plots are updated, while the scatter plots persist
 11. Continue searching for genes and toggling spatial groups, and observe that no plots go blank due to missing webGL contexts
